reactive/docker.py

```diff
@@ -919,6 +921,18 @@ def recycle_daemon():
 
     :return: None
     """
+
+    # If juju environment variables are defined, take precedent
+    # over config.yaml.
+    # See: https://github.com/dshcherb/charm-helpers/blob/eba3742de6a7023f22778ba58fbbb0ac212d2ea6/charmhelpers/core/hookenv.py#L1455
+    # &: https://bugs.launchpad.net/charm-layer-docker/+bug/1831712
+    environment_config = hookenv.env_proxy_settings()
+    configuration = config()
+    if environment_config is not None:
+        configuration.update(environment_config)
```

Comment:
The [Config object](https://github.com/juju/charm-helpers/blob/b31408ece9f39b0bc527d785acd9ab492c14ff39/charmhelpers/core/hookenv.py#L278) returned by `config()` should be considered read-only. If the charm writes to it like this, then future hooks will get incorrect results from calls to `config.previous()`, and that will cause `config.changed.http_proxy`, `config.changed.https_proxy`, and `config.changed.no_proxy` to be set on every hook execution. That will ultimately lead to `recycle_daemon` being called on every hook.